examples/hetero/recommender_system.py

```diff
@@ -0,0 +1,434 @@
+# Use only edges that have high ratings (>=4)
+rating_threshold = 4
+edge_names = [('user', 'rates', 'movie'), ('movie', 'rev_rates', 'user')]
+filter_indices = torch.where(
+    data[edge_names[0]].edge_label >= rating_threshold)[0]
+for edge_name in edge_names:
+    for attr in data[edge_name]:
+        if len(data[edge_name][attr].size()) == 1:
+            data[edge_name][attr] = data[edge_name][attr][filter_indices]
+        else:
+            data[edge_name][attr] = data[edge_name][attr][:, filter_indices]
+
+# Perform a temporal link-level split into train, val, and test edges:
+split_ratio = [0.8, 0.1, 0.1]
+assert sum(split_ratio) == 1
+_, perm = torch.sort(data[('user', 'rates', 'movie')].time)
+for edge_name in edge_names:
+    for attr in data[edge_name]:
+        if len(data[edge_name][attr].size()) == 1:
+            data[edge_name][attr] = data[edge_name][attr][perm]
+        else:
+            data[edge_name][attr] = data[edge_name][attr][:, perm]
+num_edges = len(data[edge_names[0]].time)
+
+splits = [int(i * num_edges) for i in split_ratio]
+splits[2] += num_edges - sum(splits)
+train_data = copy.deepcopy(data)
+val_data = copy.deepcopy(data)
+test_data = copy.deepcopy(data)
+graphs = [train_data, val_data, test_data]
+for edge_name in edge_names:
+    for attr in data[edge_name]:
+        dim = len(data[edge_name][attr].size()) - 1
+        data_splits = torch.split(data[edge_name][attr], splits, dim)
+        for i, graph in enumerate(graphs):
+            graph[edge_name][attr] = data_splits[i]
+
+BATCH_SIZE = 512
+
+train_dataloader = LinkNeighborLoader(
+    data=train_data, num_neighbors=[5, 5, 5], neg_sampling_ratio=1,
+    edge_label_index=(('user', 'rates', 'movie'),
+                      train_data[('user', 'rates', 'movie')].edge_index),
+    edge_label_time=train_data[('user', 'rates', 'movie')].time,
+    batch_size=BATCH_SIZE, shuffle=True, time_attr='time')
+
+val_dataloader = LinkNeighborLoader(
+    data=val_data,
+    num_neighbors=[-1, -1, -1],
+    neg_sampling_ratio=1,
+    edge_label_index=(('user', 'rates', 'movie'),
+                      val_data[('user', 'rates', 'movie')].edge_index),
+    batch_size=BATCH_SIZE,
+    shuffle=True,
+)
+
+test_dataloader = LinkNeighborLoader(
+    data=test_data,
+    num_neighbors=[-1, -1, -1],
+    neg_sampling_ratio=1,
+    edge_label_index=(('user', 'rates', 'movie'),
+                      test_data[('user', 'rates', 'movie')].edge_index),
+    batch_size=BATCH_SIZE,
+    shuffle=True,
+)
+
+# We have an unbalanced dataset with many labels for rating 3 and 4, and very
+# few for 0 and 1. Therefore we use a weighted MSE loss.
+if args.use_weighted_loss:
+    weight = torch.bincount(data['user', 'movie'].edge_label)
+    weight = weight.max() / weight
+else:
+    weight = None
+
+
+def weighted_mse_loss(pred, target, weight=None):
+    weight = 1. if weight is None else weight[target].to(pred.dtype)
+    return (weight * (pred - target.to(pred.dtype)).pow(2)).mean()
```

Comment:
Can be dropped since we only want to do link prediction, not rating prediction.

Reply:
Agreed. Removed this portion -- I'm using F.mse_loss() now. Thanks